### ISSUE TITLE
Change of git workflow: merge directly into master

### ISSUE THREAD

**Issue by mloning** [alan-turing-institute/sktime#360]
We currently follow a git workflow in which we work from the dev branch. This has a number of disadvantages:

* it adds a level of complication for new contributors, often PRs are created where new contributors have created a new branch based on dev and want to merge it into master or vice versa;
* automatic updating of the CHANGELOG does not work nicely;
* GitHub recognises only contributors that merge into master;
* PRs into dev do not trigger doc builds, so changes to tutorial notebooks are not included in CI checks;
* we need an additional PR where we collect all changes on dev and merge it into master for every new release (including merge conflicts, etc).

Instead, we could work directly based on the master branch. New releases would be triggered by tags which I have set up already. 

This is what other packages seem to do too (see e.g. scikit-learn). 

### What we would have to change:
* contributing guidelines
* automatic changelog builder (GitHub action)

**Comment by mloning** [comment]
FYI @James-Large @TonyBagnall @fkiraly @ABostrom

**Comment by fkiraly** [comment]
(I´d say this is more maintenance than governance, in terms of label?)

*I don't think it's a good idea to directly work on master*, simply since it bcomes easy to accidentally break the "current clean version".  It's best gitflow practice to have dev, and only merge to master after ensuring that things still work. Typically you would also give master a stronger branch protection than dev.

**Comment by fkiraly** [comment]
👎

**Comment by TonyBagnall** [comment]
I prefer the dev/master model, maybe just have more regular pushes from Dec to master?

**Comment by mloning** [comment]
@fkiraly 
* To clarify, we don't work directly on master, master is read-only (except for admins), people would work on separate branches, but create PR to merge directly into master rather than merging into dev
* PR runs our CI tests which ensure that everything works before we merge anything. That doesn't change. As said, the PR into master runs more comprehensive CI tests than the PR into dev. 
* If something still breaks, you can always revert to a previous commit. 
* The release process is still separate (triggered by tagging) so there's an extra level of protection for releases/distribution. 
* My current impression is that we merge things into dev thinking that it'll be reviewed again before being merged into master and then create a PR with 100+ changes for various contributions that's nearly impossible to review. My proposed workflow would still require two reviews/approvals for every PR (based on our current branch protection rules) but these reviews would be more meaningful now since they focus on particular, understandable changes rather than a collection of 100+ unrelated changes. So our current workflow may make it more likely to merge bad code. 
* Not sure what "best practice" is, many other packages seem to merge directly into master (I've checked SciPy, scikit-learn, TensorFlow). 
* PS: you have to leave some time to find a consensus and call a vote before casting your vote ;-) 

@TonyBagnall 
* In the end, we want more frequent releases and not just more pushes to master, no? Changing the workflow will make our release cycle faster, because I don't have to go through a separate PR for merging everything from dev into master (including reviews, merge conflicts, etc).

**Comment by fkiraly** [comment]
> PS: you have to leave some time to find a consensus and call a vote before casting your vote ;-)

Well, I interpreted this as discussing a concrete proposal for change, rather than "let's review our GitFlow model, any suggestions?" - also, was sticking by the voting process we wanted to introduce?

>"best practice" is, many other packages seem to merge directly into master (I've checked SciPy, scikit-learn, TensorFlow).

In the original GitFlow model
https://nvie.com/posts/a-successful-git-branching-model/

it is considered a good idea to separate things:
* consolidation of multiple features - a slightly messy endeavour if multiple people are working on a project
* updating the production release, triggering a build

It is also considered good that there is always a branch which only has stable and tested code.

Now one could consider attachment of a tag as the operation that triggers a build, but in the "reduced" model there is no branch were you can be sure that it contains only stable code.

Based on an assortment of situations in which I was tempted to send emails with subject line "don't merge while drunk", I strongly suggest to err on the side of caution here...

>Not sure what "best practice" is, many other packages seem to merge directly into master (I've checked SciPy, scikit-learn, TensorFlow).

sklearn seems to adopt the master/dev/feature GitFlow models, in the form of a feature/dev/release model?

As far as I can see:
* vanilla Gitflow dev corresponds to master on sklearn
* vanilla Gitflow master/release is the newest sklearn release branch - e.g., 0.23.0
* feature branches go into sklearn master and live on forks

Personally, I don't mind master/dev/feature or release/dev/feature, and I'm happy if we call dev master in the second.

**Comment by TonyBagnall** [comment]
since you bare the load of managing pull requests I'm inclined to do it the way you think best, but I'll discuss it with my guys

**Comment by fkiraly** [comment]
>since you bare the load of managing pull requests I'm inclined to do it the way you think best, but I'll discuss it with my guys

I very much second that, in terms of the final decision made, but it won't keep me from disagreeing based on my own opinions :-)

**Comment by mloning** [comment]
@fkiraly As far as I understand the release branches in scikit-learn, their main purpose is to have a "feature freeze". This allows to add minor changes to a given release (e.g. bug fixes) and to continue to develop major features based on master for the next release. But we don't really have this level of parallel development activity at the moment. For now, working based on master seems to best choice.

**Comment by fkiraly** [comment]
>@fkiraly As far as I understand the release branches in scikit-learn, their main purpose is to have a "feature freeze".

Yes, but it also is a suitable replacement for a "true" master branch, in capturing a consolidated version in a branch, separate from the act of merging and consolidation - so accidental destruction of the consolidated branch (in the sense of having a non-working most recent commit) is highly inlikely.

*without* the release branches, you wouldn´t have any "safe" branches - that´s why I am ok with master/dev/features, or release/dev/features, but not with master/features.

**Comment by James-Large** [comment]
There's pros and cons all round, ultimately in my opinion (which has been studied less than you guys) you want: 

- The thing that a new user pulls/installs to be first and foremost stable/working, but have as many features in it as possible
- The thing that a more interested/invested user pulls/installs to have the latest features, where there's a small chance there are bugs or the feature's api may change in the future etc. This might be dev for generically more features, or a particular branch for a particular feature they want asap

It's basically between (1):

 `master (which is also the current release version) > dev >>> features`

and (2): 

```
release(s) 
   ^
master  >>>> features
```

As @fkiraly says, master in (2) can easily just be called dev really. And agreed you don't want (2) without the stable release branches

Both make sense, and which is better I guess comes with a combination of the relative speed of updates, # of different features being developed simultaneously, likelihood of error introduction (based on familiarity+competence of coders vs complexity of codebase and coverage of tests), etc etc 

Those can't really be measured, so maybe just go with whichever sounds best to the person actually doing the maintaining

**Comment by fkiraly** [comment]
@James-Large, just to clarify my understanding: do you agree with me that neither your (1) nor (2) are the same as what @mloning originally proposes? Or, is, in your opinion, (2) what @mloning proposes (with which I would be fine)?

@mloning how many distinct options were there discussed in your opinion?

**Comment by mloning** [comment]
I'm not sure why we want to have (i) a stable release branch when working on master or (ii) work on dev and have a stable master. 

One reason is that you may want to maintain multiple versions, so you want to freeze the main features of each version but still provide bug fixes. But this is not what we're doing currently. We don't provide bug fixes previous versions, at least not for now.

Are there any other reason? 

If we work on master without any other stable branch, users/developers can still easily obtain stable versions (releases) via pip install or by checking out tags/releases in git.

**Comment by fkiraly** [comment]
@mloning, can you please comment on how many distinct options were there discussed in your opinion?
I´m not sure whether there are three or two, and whether we all agree on that number.

**Comment by mloning** [comment]
I think we have 3 options: 
* master + feature branches for development
* master + dev + feature branches for development
* master + release branches + feature branches for development

**Comment by fkiraly** [comment]
>I think we have 3 options:

Ok, that´s what I thought.
I´m ok with no.2 and no.3, but strongly recommend against no.1.

**Comment by mloning** [comment]
Can you explain why you're against (1) (see my previous post)?

**Comment by fkiraly** [comment]
>Can you explain why you're against (1) (see my previous post)?

Sure.

Premises:

Premise 1 - there are "types of branches" in Git, including the type "feature branch". Each branch has a type. Why: it's a typical concept in a GitFlow model.
Premise 2 - there must be a branch of type "stable branch", in the sense that at any time, the most recent commit is well-tested and reliable, which counts as a "stable" package version. Why: having a stable release quickly available for developers is considered important, and the common convention is to map this onto a branch (as opposed to "repo specific conventions").
Premise 3 - branches into which "feature branches" are directly merged cannot be "stable branches". Why: the merging process is such that bugs occur on occasion. Consolidation and testing has to occur before a post-merge commit can be considered "stable" in the above sense.

Main consequence: if "feature branches" content is to be ultimatly merged into a "stable branch", there need to be at least 3 branch types in the GitFlow model.
Proof: By premise 2, a "stable branch" exists. .By premise 3, any branch type that a "feature branch" is directly merged into is not a "stable branch". Since feature content needs to be merged into it by the assumption in the statement, it can therefore happen only indirectly, i.e., via a sequence of branches at the end of which is a "stable branch", all of which have a type by premise 1. So we can consider the last branch in the sequencce whose type is not a "stable branch" - this exists since the sequence start isn't, and the sequence end is. By construction, it is not a "stable branch". By premise 2, it is not a "feature branch". Thus the statement consequene is proven.

**Comment by mloning** [comment]
Premise 2 :)

**Comment by fkiraly** [comment]
>why follow the convention when we don't gain anything from it and it

I don't agree that nothing is gained from this convention!

Key benefits:
* developers can easily find a stable version to start off and clone. Without a stable branch, this is not possible, or more difficult. A common convention is that "it's always safe to deploy master".
* it becomes easier to revert development to a stable version, or use the last stable version for a merge or rebase. Otherwise, you would have to search for the last commit considered stable.
* when not all reviewers are very experienced, it creates an additional "firewall" against bugs, merge problems, etc. Merge conflicts should approach master no closer than dev!

Further, the main complication of yours I don't really see.
Dev is always supposed to be ahead of master, so once you reach a state in dev that is considered "safe" you, PR to master.

Btw, Github's GitFlow is a simplified version which has two branches (no develop):
https://githubflow.github.io/
But the premise they disagree with is premise 3, not premise 2! I would say it is rare and uncommon to disagree with premise 2.
Based on your disagreement with premise 2, I would make my recommendation stronger.

Regarding "maturity" - I assume you imply it's low? Or high? In either case, I'd rather base it on the developer base experience, and I expect we have the full spectrum. Which in my opinion speaks rather for having more watersheds than less (so people can more "safely" mess up accidentally, if it comes to that).

**Comment by mloning** [comment]
I've commented on the key benefits and complications above ...

**Comment by fkiraly** [comment]
>I've commented on the key benefits and complications above ...

Where exactly, can you please point me?
Is this new material, or existing material?

Just to summarize our discussion so far:
* you agree with premise 3, disagree with premise 2
* I think premise 2 is very important

Agreed?

Do we also agree that premise 2 is held as very important by "high profile" dev teams, such as Github's dev team?

**Comment by fkiraly** [comment]
In terms of how we proceed, any good suggestions? 
@James-Large @TonyBagnall @ABostrom @mloning
We seem to disagree on "axioms" or "beliefs", if you will.
 Some options I can think of:
* I do usually not like the "argumentum ad verecundiam" (authority arguments), but we could try to collate high profile projects or teams that have made statements about the importane of premise 2 or its opposite.
* We could also ask our developer level users what they think?

**Comment by TonyBagnall** [comment]
I think (3) is probably the best, although I do not feel strongly. I think release branches are sensible, and I intend having "paper branches" to make sure code/results we quote in papers are reproducible in the future

**Comment by TonyBagnall** [comment]
from my perspective, the problem with (2) is mainly perception, (everyone sees master by default) and the fact that dev->master ends up not happening very often. I guess the difference between (1) and (2) is what is meant by a release. I'm not really sure how it works. If we are looking at continuous release (each update to master is a new release) then a separate branch for each will just get bloaty. If, however, releases are distinct to Master PRs, and happen less frequently, then (3) sounds best.

**Comment by fkiraly** [comment]
>rom my perspective, the problem with (2) is mainly perception, (everyone sees master by default) 

Hm, if that's a problem, that's easily fixable, for example by:
* setting default branch to `dev`, then everyone sees `dev` by default, and PR go there by default. Settings -> Branches -> Default branch. That's an easy convenience fix which I often have in my private repos.
* using model (3), in which the "dev-like" branch is called master

**Comment by fkiraly** [comment]
@TonyBagnall, I think paper branches are an interesting idea. An alternative might be use of tags.

Though that's another class of flow models - or at least an extension of either model - that we would be looking at.

Shall we discuss in our next regular TC meeting, which I guess is this Fri?

**Comment by TonyBagnall** [comment]
@fkiraly sure, happy to discuss

**Comment by mloning** [comment]
I've changed the workflow now and updated the documentation (#364). 

* We will work based on master (new feature branches fork off master, PRs get merged into master)
* Stable version are available as tags (e.g. `git checkout v0.4.1`)
